### PR TITLE
[AIRFLOW-3044] Dataflow operators accept templated job_name param

### PR DESCRIPTION
**Pull request by jeffkpayne**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3044
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
See https://issues.apache.org/jira/browse/AIRFLOW-3044

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Updated `tests.contrib.hooks.test_gcp_dataflow_hook.*` and `tests.contrib.operators.test_dataflow_operator.*`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by jeffkpayne** [comment]
Fixing now...

File: tests/contrib/hooks/test_gcp_dataflow_hook.py [hunk 1]
```diff
@@ -178,18 +178,18 @@ def poll_resp_error():
...
     def test_valid_dataflow_job_name(self):
         job_name = self.dataflow_hook._build_dataflow_job_name(
-            task_id=TASK_ID, append_job_name=False
+            job_name=JOB_NAME, append_job_name=False
         )
 
-        self.assertEquals(job_name, TASK_ID)
+        self.assertEquals(job_name, JOB_NAME)
 
     def test_fix_underscore_in_task_id(self):
```
Comment (kaxil) [inline tests/contrib/hooks/test_gcp_dataflow_hook.py]: `test_fix_underscore_in_task_id` -> you might want to rename the name of this method and its associated test

File: airflow/contrib/hooks/gcp_dataflow_hook.py [hunk 2]
```diff
@@ -208,56 +206,54 @@ def _set_variables(variables):
...
+        if not re.match(r"^[a-z]([-a-z0-9]*[a-z0-9])?$", base_job_name):
             raise ValueError(
                 'Invalid job_name ({}); the name must consist of'
                 'only the characters [-a-z0-9], starting with a '
-                'letter and ending with a letter or number '.format(task_id))
+                'letter and ending with a letter or number '.format(base_job_name))
 
         if append_job_name:
-            job_name = task_id + "-" + str(uuid.uuid4())[:8]
+            safe_job_name = base_job_name + "-" + str(uuid.uuid4())[:8]
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_dataflow_hook.py:249]: Do we need appending if a user specifies job_name 🤔 ?

**Review by kaxil — CHANGES_REQUESTED** [review]
It would be ideal if we can get some input from @fenglu-g @tswast as this contains backwards-incompatible changes as the parameter positions are changed.

File: tests/contrib/hooks/test_gcp_dataflow_hook.py (hunk 1 above)
Comment (jeffkpayne) [reply tests/contrib/hooks/test_gcp_dataflow_hook.py]: Gah, yes...

File: airflow/contrib/hooks/gcp_dataflow_hook.py (hunk 2 above)
Comment (jeffkpayne) [reply airflow/contrib/hooks/gcp_dataflow_hook.py:249]: Good question.  Provided job name could still be non-unique, but yeah, could be handled by the caller in that case.  Would still be useful during testing, IMO.

File: tests/contrib/hooks/test_gcp_dataflow_hook.py (hunk 1 above)
Comment (jeffkpayne) [reply tests/contrib/hooks/test_gcp_dataflow_hook.py]: Done.

**Comment by jeffkpayne** [comment]
Fixing conflicts now...

**Comment by tswast** [comment]
> backwards-incompatible changes as the parameter positions are changed.

Do we care about the order of keyword arguments. In the google-cloud-python libraries, we consider the order of keyword arguments not relevant for determining if something is a breaking change. (Even though in Python someone could rely on the order and pass in arguments without keywords, they probably shouldn't.)

**Comment by kaxil** [comment]
@tswast This PR also changes positional arguments for methods in hooks. And if you see most of the operators use positional arguments and not keyword arguments. An example is change in `_start_dataflow` method which is then used by `start_java_dataflow` based on position and not keyword.

Would appreciate if you can review the entire PR if you have time :)

**Review by kaxil — APPROVED** [review]
LGTM

**Pull request merged**